### PR TITLE
Use run tool in coreclr dev workflow.

### PR DESCRIPTION
**Pull request by Priya91**
These changes alter the dev workflow that's been in use so far. This is a first step to move towards consuming the run tool to synchronize the dev workflow across the dotnet repos. For more information, please refer to this documentation [here](https://github.com/dotnet/buildtools/blob/master/Documentation/RunCommand.md)

cc @maririos @markwilkie @weshaggard @gkhanna79 @jashook @wtgodbe @jaredpar @sokket @joshfree 

The PRs are expected to FAIL, since the groovy script gets updated only after merge. True validation can happen only after groovy change is pushed along with the scripts. I have performed local testing on Ubuntu16.04 and Windows10 of each of the codepaths the scripts can execute.

### CODE REVIEW

**Comment by Priya91** [comment]
@dotnet-bot test ci please

File: config.json [hunk 1]
```diff
@@ -0,0 +1,541 @@
...
+          "MsBuildParameters": "default",
+          "maxcpucount": "default",
+          "MsBuildFileLogging": "/flp:v=detailed;LogFile=sync.log"
+        }
+      }
+    },
+    "publish-packages": {
+      "alias": {
+        "azureAccount": {
+          "description": "Account token to connect to Azure Blob storage.",
```
Comment (maririos) [inline config.json]: The `description` and `setting` doesn't match the name of the alias. `AzureAccount != CloudDropAccessToken`.
Same in sync with this two properties :)

File: config.json [hunk 2]
```diff
@@ -0,0 +1,541 @@
...
+        "toolName": "msbuild",
+        "settings": {
+          "Project": "./src/publish.proj",
+          "MsBuildParameters": "default",
+          "maxcpucount": "default",
+          "MsBuildFileLogging": "/flp:v=detailed;LogFile=publish-packages.log"
+        }
+      }
+    },
+    "build-packages": {
```
Comment (maririos) [inline config.json]: What is the `Project` here? is it different than when you call `build buildPackages`?

File: config.json [hunk 3]
```diff
@@ -0,0 +1,541 @@
...
+            "__BuildArch": "default"
+          }
+        },
+        "buildType": {
+          "description": "Specifies configuration to publish, can be Release, Debug or Checked.",
+          "settings": {
+            "__BuildType": "default"
+          }
+        },
+        "distroRid": {
```
Comment (maririos) [inline config.json]: For now we could try to unify common alias names between the repos. I'm using `osversion` for your `distroRid` and I use `os` for your `buildOS`.
PR dotnet/corefx#10231

File: config.json [hunk 4]
```diff
@@ -0,0 +1,541 @@
...
+      "values": [ true, false ],
+      "defaultValue": true
+    },
+    "Project": {
+      "description": "Project where the commands are going to be applied.",
+      "valueType": "passThrough",
+      "values": [],
+      "defaultValue": ""
+    },
+    "__BuildType": {
```
Comment (maririos) [inline config.json]: In this context, what is the difference between the Settings that start with `__` and the ones that don't? Is it a convention we want to maintain?

File: build-packages.cmd [hunk 5]
```diff
@@ -2,61 +2,43 @@
...
-  echo ERROR: An error occurred while building packages, see %packagesLog% for more details.
+  echo ERROR: An error occurred while building CoreCLR Runtime package, see build-packages.log for more details.
   exit /b 1
 )
 
-set __msbuildArgs="%__ProjectDir%\src\.nuget\Microsoft.NETCore.Jit\Microsoft.NETCore.Jit.builds" !allargs!
-echo msbuild.exe %__msbuildArgs% !options! >> %packagesLog%
-call msbuild.exe %__msbuildArgs% !options!
+echo %__ProjectDir%/run.cmd build-packages -Project=%__ProjectDir%/src/.nuget/Microsoft.NETCore.Jit/Microsoft.NETCore.Jit.builds -FilterToOSGroup=Windows_NT %allargs%
+call %__ProjectDir%/run.cmd build-packages -Project=%__ProjectDir%/src/.nuget/Microsoft.NETCore.Jit/Microsoft.NETCore.Jit.builds -FilterToOSGroup=Windows_NT %allargs%
```
Comment (maririos) [inline build-packages.cmd]: You could create an alias for `FilterToOSGroup` and then pass default. This is going to run in windows so run.exe should set the value to `Windows_NT` and that way you don't need to sent it in every call.

File: build-packages.cmd [hunk 6]
```diff
@@ -2,61 +2,43 @@
...
 if NOT [!ERRORLEVEL!]==[0] (
-  echo ERROR: An error occurred while building packages, see %packagesLog% for more details.
+  echo ERROR: An error occurred while building CoreCLR Runtime package, see build-packages.log for more details.
   exit /b 1
 )
 
-set __msbuildArgs="%__ProjectDir%\src\.nuget\Microsoft.NETCore.Jit\Microsoft.NETCore.Jit.builds" !allargs!
-echo msbuild.exe %__msbuildArgs% !options! >> %packagesLog%
-call msbuild.exe %__msbuildArgs% !options!
+echo %__ProjectDir%/run.cmd build-packages -Project=%__ProjectDir%/src/.nuget/Microsoft.NETCore.Jit/Microsoft.NETCore.Jit.builds -FilterToOSGroup=Windows_NT %allargs%
```
Comment (maririos) [inline build-packages.cmd]: I think run.cmd is printing the command before running it. You could delete the echos in here

File: build-packages.sh [hunk 7]
```diff
@@ -3,159 +3,65 @@
...
-            ;;
-        debug)
-            __BuildType=Debug
-            ;;
-        release)
-            __BuildType=Release
-            ;;
-        checked)
-            __BuildType=Checked
+        usage
```
Comment (maririos) [inline build-packages.sh]: Is the idea of the loop to get rid of the parameter `-?` I don't see what else are you taking out of `$*`. 
If this is the case you could ask 
```
if [ $1 == "-?" ]; then
    usage
fi
```
And then just pass `$*` to the project calls

File: build-packages.cmd [hunk 8]
```diff
@@ -67,7 +49,7 @@ exit /b
 echo.
```
Comment (maririos) [inline build-packages.cmd]: For consistency we should decide on what should happen when the user types `-?`.
In PR dotnet/corefx#10231 I'm using the documentation that run.exe provides for the specific command which is one of the purposes of it, but sometimes I'm printing both, the one in the script and the one in run.exe.
The one in the script is to help developers understand what is happening and how the scripts are changing, like for example: [build.cmd](https://github.com/maririos/corefx/blob/56e4903a9c161e27e814366ff40ae3b1f93901d0/build.cmd#L12)

File: build.cmd [hunk 9]
```diff
@@ -189,429 +148,256 @@ set "__TestRootDir=%__RootBinDir%\tests"
...
 
-if not defined VSINSTALLDIR (
-    echo %__MsgPrefix%Error: VSINSTALLDIR variable not defined.
-    exit /b 1
-)
-if not exist "%VSINSTALLDIR%DIA SDK" goto NoDIA
+    :: Set the environment for the native build
+    set __VCBuildArch=x86_amd64
+    if /i "%__BuildArch%" == "x86" ( set __VCBuildArch=x86 )
+    echo %__MsgPrefix%Using environment: "%VS140COMNTOOLS%\..\..\VC\vcvarsall.bat" %__VCBuildArch%
```
Comment (janvorli) [inline build.cmd]: Why have you replaced the __VSToolsRoot with VS140COMNTOOLS everywhere? Now when we will change the version of tools we use in the future, it will need to be modified at 17 different places instead of just one.

File: build.cmd [hunk 10]
```diff
@@ -189,429 +148,256 @@ set "__TestRootDir=%__RootBinDir%\tests"
...
-set "__CrossGenCoreLibLog=%__LogsDir%\CrossgenMSCoreLib_%__BuildOS%__%__BuildArch%__%__BuildType%.log"
-set "__CrossgenExe=%__CrossComponentBinDir%\crossgen.exe"
-"%__CrossgenExe%" /Platform_Assemblies_Paths "%__BinDir%" /out "%__BinDir%\mscorlib.ni.dll" "%__BinDir%\mscorlib.dll" > "%__CrossGenCoreLibLog%" 2>&1
-if NOT %errorlevel% == 0 (
-    echo %__MsgPrefix%Error: CrossGen mscorlib facade build failed. Refer to the build log file for details:
-    echo     %__CrossGenCoreLibLog%
-    type     %__CrossGenCoreLibLog%
-    exit /b 1
-)
+    set __msbuildLog=/flp:Verbosity=normal;LogFile="%__LogsDir%\Nuget_%__BuildOS%__%__BuildArch%__%__BuildType%.log"
```
Comment (janvorli) [inline build.cmd]: Could you please use the same naming convention for all variables? Before your change, the variable names started with two underscores and then an uppercase letter. Now some of them start with single underscore and some of them with lower case letter after the underscore.

File: build.sh [hunk 11]
```diff
@@ -42,9 +40,15 @@ usage()
     echo "skipnative - do not build native components."
     echo "skipmscorlib - do not build mscorlib.dll."
     echo "skiptests - skip the tests in the 'tests' subdirectory."
-    echo "skiprestore - skip restoring nuget packages."
     echo "skipnuget - skip building nuget packages."
-    echo "disableoss - Disable Open Source Signing for mscorlib."
+    echo "verbose - optional argument to enable verbose build output."
+    echo "-skiprestore: skip restoring packages ^(default: packages are restored during build^)."
```
Comment (janvorli) [inline build.sh]: I assume the dashes at the options below are just typos, right?

File: build.sh [hunk 12]
```diff
@@ -30,8 +30,6 @@ usage()
     echo "Usage: $0 [BuildArch] [BuildType] [clean] [verbose] [coverage] [cross] [clangx.y] [ninja] [configureonly] [skipconfigure] [skipnative] [skipmscorlib] [skiptests] [cmakeargs] [bindir]"
     echo "BuildArch can be: x64, x86, arm, arm-softfp, arm64"
     echo "BuildType can be: debug, checked, release"
-    echo "clean - optional argument to force a clean build."
```
Comment (janvorli) [inline build.sh]: Is the removal of the "clean" option intentional?

File: config.json (hunk 1 above)
Comment (Priya91) [inline config.json]: The alias azureaccount, sets the setting clouddropaccesstoken.. Not sure what you meant..

File: config.json (hunk 2 above)
Comment (Priya91) [inline config.json]: yea, this alias builds the various nuspecs, the projects are provided in build-packages.cmd..

File: config.json (hunk 3 above)
Comment (Priya91) [inline config.json]: distroRid is better than osversion, since this provides the full runtime id, for ex, this value will be ubuntu.16.04-x64, where osversion will be ubuntu.16.04. And using buildOS here since that's the name of setting used through the coreclr build targets.

File: config.json (hunk 4 above)
Comment (Priya91) [inline config.json]: the __ is the setting that sets the properties used in coreclr targets, the ones without are aliases not settings..

File: build-packages.cmd (hunk 5 above)
Comment (Priya91) [inline build-packages.cmd]: FilterToOSGroup is not an alias for build-packages.sh.. so didn't want to provide it as an alias, since it's not common to both..

File: build.sh (hunk 11 above)
Comment (Priya91) [inline build.sh]: No, there are the options that will not be directly passed to run command, which will set the necessary msbuild properties based on the information provided from config.json.. The idea is to move most of these options into run, so not much parsing has to happen in build.cmd..

File: build.sh (hunk 12 above)
Comment (Priya91) [inline build.sh]: Yes, clean.cmd does the same thing..

File: build.cmd (hunk 9 above)
Comment (Priya91) [inline build.cmd]: Will change this

File: build.cmd (hunk 10 above)
Comment (Priya91) [inline build.cmd]: Will do

File: build-packages.sh (hunk 7 above)
Comment (Priya91) [inline build-packages.sh]: Will do

File: build-packages.cmd (hunk 6 above)
Comment (Priya91) [inline build-packages.cmd]: Will do

File: config.json [hunk 13]
```diff
@@ -0,0 +1,541 @@
+{
+  "settings": {
+    "CloudDropAccessToken": {
```
Comment (jaredpar) [inline config.json]: Are these settings relevant to every command or just specific ones?

File: config.json (hunk 1 above)
Comment (maririos) [inline config.json]: azureAccount is setting the property CloudDropAccessToken . It should set CloudDropAccountName, it makes more sense.
and then azureToken sets CloudDropAccessToken

File: config.json (hunk 2 above)
Comment (maririos) [inline config.json]: Is there a way to make this more clear? it is a little confusing

File: config.json (hunk 3 above)
Comment (maririos) [inline config.json]: I'll change mine then.

File: config.json (hunk 1 above)
Comment (Priya91) [inline config.json]: ohh, i reversed the alias names :) good catch!

File: config.json (hunk 2 above)
Comment (Priya91) [inline config.json]: There are multiple projects to build, not sure how to associate this command for multiple projects..

File: config.json (hunk 13 above)
Comment (Priya91) [inline config.json]: Relevant only to the publish-packages command.. But without specifying in the global settings, run tool throws error that the setting is not found..

**Comment by Priya91** [comment]
@dotnet-bot test ci please

**Comment by Priya91** [comment]
@dotnet-bot test ci please

**Comment by Priya91** [comment]
@dotnet-bot test Linux ARM Emulator Cross Debug Build

**Comment by Priya91** [comment]
@dotnet-bot test ci please

**Comment by Priya91** [comment]
The windows priority job is failing with wrong parameters to run, should be fixed with the netci changes upon merge. The Checked jobs are failures are not related to this PR, they are failing with same error in other PRs as well.

**Comment by Priya91** [comment]
@gkhanna79 @wtgodbe Can you please review the changes..

**Comment by Priya91** [comment]
@dotnet-bot test Linux ARM Emulator Cross Debug Build

**Comment by Priya91** [comment]
@dotnet-bot test Linux ARM Emulator Cross Release Build

**Comment by Priya91** [comment]
@maririos @weshaggard Can you please review the changes..

**Comment by maririos** [comment]
@dotnet-bot test OSX x64 Checked Build and Test please

**Comment by maririos** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test please

**Comment by gkhanna79** [comment]
@wtgodbe Can you PTAL?

File: build.sh [hunk 14]
```diff
@@ -42,9 +40,15 @@ usage()
     echo "skipnative - do not build native components."
     echo "skipmscorlib - do not build mscorlib.dll."
     echo "skiptests - skip the tests in the 'tests' subdirectory."
-    echo "skiprestore - skip restoring nuget packages."
     echo "skipnuget - skip building nuget packages."
-    echo "disableoss - Disable Open Source Signing for mscorlib."
+    echo "verbose - optional argument to enable verbose build output."
+    echo "-skiprestore: skip restoring packages ^(default: packages are restored during build^)."
+	echo "-disableoss: Disable Open Source Signing for System.Private.CoreLib."
+	echo "-priority=^<N^> : specify a set of test that will be built and run, with priority N."
```
Comment (wtgodbe) [inline build.sh]: I don't believe you need the 'priority' arg here, since we don't build tests on non-windows yet

File: netci.groovy [hunk 15]
```diff
@@ -1662,10 +1662,10 @@ combinedScenarios.each { scenario ->
 
                                     // Debug runs take too long to run. So build job only.
                                     if (lowerConfiguration == "debug") {
-                                       buildCommands += "set __TestIntermediateDir=int&&build.cmd ${lowerConfiguration} ${architecture} /toolset_dir C:\\ats2"
+                                       buildCommands += "set __TestIntermediateDir=int&&build.cmd ${lowerConfiguration} ${architecture} toolset_dir C:\\ats2"
                                     }
                                     else {
-                                       buildCommands += "set __TestIntermediateDir=int&&build.cmd skiptests ${lowerConfiguration} ${architecture} /toolset_dir C:\\ats2"
+                                       buildCommands += "set __TestIntermediateDir=int&&build.cmd skiptests ${lowerConfiguration} ${architecture} toolset_dir C:\\ats2"
```
Comment (wtgodbe) [inline netci.groovy]: Do we need to do anything different for the invocations of build.sh in this file?

**Comment by wtgodbe** [comment]
LGTM, other than 2 comments

**Comment by maririos** [comment]
I would like to call `-?` directly from the run tool and not from the scripts. For the places where we can, we should do that.
Otherwise, LGTM

File: netci.groovy (hunk 15 above)
Comment (Priya91) [inline netci.groovy]: No, most of the build.sh commands are parsed in build.sh right now..

**Comment by Priya91** [comment]
@maririos Will file an issue to track it.. some scripts, for, ex build.sh/build.cmd, still have some parameters that are not passed to run, will tackle this in a separate PR.

File: build.sh (hunk 14 above)
Comment (Priya91) [inline build.sh]: Removed..

**Comment by Priya91** [comment]
@dotnet-bot test ci please

File: build.cmd [hunk 16]
```diff
@@ -189,429 +148,256 @@ set "__TestRootDir=%__RootBinDir%\tests"
...
+	echo %__MsgPrefix%Commencing build of System.Private.CoreLib for %__BuildOS%.%__BuildArch%.%__BuildType%
+    rem Explicitly set Platform causes conflicts in CoreLib project files. Clear it to allow building from VS x64 Native Tools Command Prompt
+    set Platform=
+
+    :: Set the environment for the managed build
+    echo %__MsgPrefix%Using environment: "%__VSToolsRoot%\VsDevCmd.bat"
+    call                                 "%__VSToolsRoot%\VsDevCmd.bat"
+
+    set __MsbuildLog=/flp:Verbosity=normal;LogFile="%__LogsDir%\System.Private.CoreLib_%__BuildOS%__%__BuildArch%__%__BuildType%.log"
+	set __MsbuildWrn=/flp1:WarningsOnly;LogFile="%__LogsDir%\System.Private.CoreLib_%__BuildOS%__%__BuildArch%__%__BuildType%.wrn"
```
Comment (weshaggard) [inline build.cmd]: nit: some of these lines are indented and some are not.

File: build-packages.sh [hunk 17]
```diff
@@ -3,159 +3,64 @@
...
-    # Invoke MSBuild
-    $__ProjectRoot/Tools/dotnetcli/dotnet "$__MSBuildPath" /nologo "$__ProjectRoot/src/.nuget/Microsoft.NETCore.Runtime.CoreCLR/Microsoft.NETCore.Runtime.CoreCLR.builds" /verbosity:minimal "/fileloggerparameters:Verbosity=normal;LogFile=$binclashlog" /t:Build /p:__BuildOS=$__BuildOS /p:__BuildArch=$__BuildArch /p:__BuildType=$__BuildType /p:__IntermediatesDir=$__IntermediatesDir /p:BuildNugetPackage=false /p:UseSharedCompilation=false
-
-if [ $? -ne 0 ]; then
-    echo -e "\nAn error occurred. Aborting build-packages.sh ." >> $build_packages_log
-    echo "ERROR: An error occurred while building packages, see $build_packages_log for more details."
+$__ProjectRoot/run.sh build-packages -Project=$__ProjectRoot/src/.nuget/Microsoft.NETCore.Runtime.CoreCLR/Microsoft.NETCore.Runtime.CoreCLR.builds -DistroRid=\${OSRid}-$__Arch -UseSharedCompilation=false -BuildNugetPackage=false $unprocessedBuildArgs
+if [ $? -ne 0 ]
+then
+    echo "ERROR: An error occurred while syncing packages; See build-packages.log for more details."
```
Comment (sergiy-k) [inline build-packages.sh]: Should the error message say "while building packages" instead of "syncing"? Same in a few places below.

File: build.sh [hunk 18]
```diff
@@ -176,13 +162,14 @@ build_coreclr()
...
-            echo "Running: \"$__ProjectRoot/Tools/dotnetcli/dotnet\" \"$__ProjectRoot/Tools/MSBuild.exe\" \"$__ProjectRoot/build.proj\" /v:minimal /t:GenerateVersionSourceFile /p:NativeVersionSourceFile=$__versionSourceFile /p:GenerateVersionSourceFile=true /v:minimal $__OfficialBuildIdArg"
-            "$__ProjectRoot/Tools/dotnetcli/dotnet" "$__ProjectRoot/Tools/MSBuild.exe" "$__ProjectRoot/build.proj" /v:minimal /t:GenerateVersionSourceFile /p:NativeVersionSourceFile=$__versionSourceFile /p:GenerateVersionSourceFile=true /v:minimal $__OfficialBuildIdArg
+            echo "Running: \"$__ProjectRoot/run.sh\" build -generateHeaderUnix -NativeVersionSourceFile=$__versionSourceFile $__RunArgs $__UnprocessedBuildArgs"
+            "$__ProjectRoot/run.sh" build -generateHeaderUnix -NativeVersionSourceFile=$__versionSourceFile $__RunArgs $__UnprocessedBuildArgs
         else
             __versionSourceLine="static char sccsid[] __attribute__((used)) = \"@(#)No version information produced\";"
             echo $__versionSourceLine > $__versionSourceFile
         fi
+
+		pushd "$__IntermediatesDir"
```
Comment (sergiy-k) [inline build.sh]: Changing the current directory to $__IntermediatesDir" here (instead of where it was done previously) will break the build if __SkipConfigure is set. The make command must be run from $__IntermediatesDir".

File: build.sh [hunk 19]
```diff
@@ -215,16 +202,7 @@ build_coreclr()
...
-
-restoreBuildTools()
-{
-    echo "Restoring BuildTools..."
-    $__ProjectRoot/init-tools.sh
-    if [ $? -ne 0 ]; then
-        echo "Failed to restore BuildTools."
-        exit 1
-    fi
+	popd
```
Comment (sergiy-k) [inline build.sh]: This "popd" command will not match the "pushd" command above if __SkipConfigure is set.

**Comment by sergiy-k** [comment]
@Priya91, thank you! LGTM modulo comments.

File: build.sh (hunk 18 above)
Comment (Priya91) [inline build.sh]: The run tool needs to run from root, i pushed and poped again later for running the make tool.

File: build.sh (hunk 19 above)
Comment (Priya91) [inline build.sh]: Fixed, with an extra push outside of the if statement.

**Comment by Priya91** [comment]
@dotnet-bot test ci please

**Comment by Priya91** [comment]
@dotnet-bot test ci please

**Comment by Priya91** [comment]
@dotnet-bot test Linux ARM Emulator Cross Debug Build

**Comment by Priya91** [comment]
@dotnet-bot test Windows_NT x64 Debug Build and Test

**Comment by Priya91** [comment]
@dotnet-bot test Windows_NT x86 ryujit Checked Build and Test
